internal/k8s/ingressstatus.go

```diff
@@ -14,45 +14,94 @@
 package k8s
 
 import (
+	"github.com/projectcontour/contour/internal/annotation"
 	"github.com/sirupsen/logrus"
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/api/networking/v1beta1"
 	clientset "k8s.io/client-go/kubernetes"
 )
 
-// StatusLoadbalancerUpdater observes informer OnAdd events and
+// IngressStatusUpdater observes informer OnAdd events and
 // updates the ingress.status.loadBalancer field on all Ingress
 // objects that match the ingress class (if used).
 type IngressStatusUpdater struct {
-	Client clientset.Interface
-	Logger logrus.FieldLogger
-	Status v1.LoadBalancerStatus
+	Client       clientset.Interface
+	Logger       logrus.FieldLogger
+	Status       v1.LoadBalancerStatus
+	IngressClass string
 }
 
 func (s *IngressStatusUpdater) OnAdd(obj interface{}) {
-	ing := obj.(*v1beta1.Ingress).DeepCopy()
 
-	// TODO(dfc) check ingress class
+	ing := obj.(*v1beta1.Ingress).DeepCopy()
+	if !annotation.MatchesIngressClass(ing, s.IngressClass) {
+		s.Logger.
+			WithField("name", ing.GetName()).
+			WithField("namespace", ing.GetNamespace()).
+			WithField("ingress-class", annotation.IngressClass(ing)).
+			Debug("Unmatched ingress class, skip status update")
+		return
+	}
 
 	ing.Status.LoadBalancer = s.Status
 	_, err := s.Client.NetworkingV1beta1().Ingresses(ing.GetNamespace()).UpdateStatus(ing)
 	if err != nil {
 		s.Logger.
 			WithField("name", ing.GetName()).
 			WithField("namespace", ing.GetNamespace()).
-			WithError(err).Error("unable to update status")
+			WithError(err).Error("Unable to update status")
 	}
 }
 
 func (s *IngressStatusUpdater) OnUpdate(oldObj, newObj interface{}) {
-	// Ignoring OnUpdate allows us to avoid the message generated
-	// from the status update.
-
-	// TODO(dfc) handle these cases:
-	// - OnUpdate transitions from an ingress class which is out of scope
-	// to one in scope.
-	// - OnUpdate transitions from an ingress class in scope to one out
-	// of scope.
+
+	oldIng := oldObj.(*v1beta1.Ingress).DeepCopy()
+	newIng := newObj.(*v1beta1.Ingress).DeepCopy()
+
+	// It's important that we only act when we are sure that there has been a change in the
+	// annotation between the old and new versions, otherwise we could generate a status update
+	// in response to another status update.
+	if !annotation.MatchesIngressClass(newIng, s.IngressClass) {
+		// new version of Ingress does not match ingress class, we may need
+		// to clear the status.
+		if annotation.MatchesIngressClass(oldIng, s.IngressClass) {
+			// The old version did match the ingress class, this object has
+			// fallen out of scope. Reset the status info.
+			s.Logger.
+				WithField("name", newIng.GetName()).
+				WithField("namespace", newIng.GetNamespace()).
+				WithField("ingress-class", annotation.IngressClass(newIng)).
+				Debug("Updated Ingress is out of scope, clearing status info")
+			newIng.Status.LoadBalancer = v1.LoadBalancerStatus{}
```

Comment:
I wonder about this. If we are no longer responsible for this object, should we touch it at all? Maybe some other ingress controller is updating the status and we could be clearing that value? There's a a number of race conditions around this ...

Reply:
Yeah, you're right. In the event that it falls out of our scope, it's not ours anymore, and we shouldn't touch it. So this function really should only handle things coming into our scope. I'll update it to that effect, with a comment to explain.